FC-1145: :memo:  Foundations Chapter (#8)
Templum/developers-community: pages/Documents/DeveloperTools/LivePersonFunctions/Foundations/external-invocation.md

---
pagename: External Invocation
keywords:
sitesection: Documents
categoryname: Developer Tools
documentname: LivePerson Functions
subfoldername: Foundations
permalink: liveperson-functions-foundations-external-invocation.html
indicator: both
---

In this chapter we will focus a bit more on the ability to directly call a functions, which is also refer to as external invocation. This gives the users of our platform the ability to [trigger](liveperson-functions-foundations-concepts.html#triggers) there functions externally through their own systems.

### Authentication

For invocation flows we allow the usage of bearer from an users, given he has the [necessary privileges](liveperson-functions-permission-system.html), or OAuth 2.0. LivePerson services may leverage additionally OAuth 1.0.

> We support the following [grant types](https://oauth.net/2/grant-types/) out of the OAuth 2.0 specification. If you want to learn more about OAuth 2.0 check out [this video](https://www.youtube.com/watch?v=CPbvxxslDTU):

Author: It was already used in the existing documentation. Hence I believe it is fine. Actually will be finally on our team lead to call a shot. Will make a not for that 👍

> We support the following [grant types](https://oauth.net/2/grant-types/) out of the OAuth 2.0 specification. If you want to learn more about OAuth 2.0 check out [this video](https://www.youtube.com/watch?v=CPbvxxslDTU):

Author: Spoke with him, he is fine with it

### API

In order to interact with our systems we are providing libraries for javascript and java. Those allow you interact with our API in a stable and secure way. Within this chapter we will explore how to leverage them and what functionalities they offer. Make also sure to checkout our [Error Codes & HTTP Responses](liveperson-functions-foundations-error-codes.html).

#### Java Client

Our client is available within Maven Central. You can head over [here](https://mvnrepository.com/artifact/com.liveperson.faas/functions-client) to get the latest version. Than you can add it to your existing setup using the below xml snippet:

```xml
<dependency>
            <groupId>com.liveperson.faas</groupId>
            <artifactId>Functions-client</artifactId>
            <version>x.x.x.x</version>
</dependency>
```

{:.notice}
The default Authentication method is OAuth 2.0 with grant type `client_credentials`. Please check the [above chapter](#authorization-process) on the necessary steps

You can leverage the following method that are exposed by the client:

> * `invokeByUUID` this method allows to invoke a productive function by it's UUID. You may leverage `isImplemented` to check if the function is productive

Author: Wanted to stay as close as possible to the actual use. In order to avoid any potential friction here.

* `invokeByEvent` this method allows to invoke all productive functions listening to the specified event. Make sure to leverage `isImplemented` to check if the event is implemented
* `isImplemented` this method allows to check if an event or a specific uuid is implemented. It is a best practice to check this prior to invoking
* `getLambdas` this method will return all functions and there states that belong to specified account

{:.important}
A more detailed version of the documentation can be found inside our [github repository](https://github.com/LivePersonInc/faas-client-jdk). If you experience an issue or have a wish, please create a github issue.

**Error Handling:**

LivePerson Functions can raise different kind of exceptions. General exceptions are wrapped in a `FaaSException`.

> `FaaSDetailedExceptions` occur when the request has been rejected by the service. For instance if the UUID of the lambda for invocation request does not exist.

Reviewer: ```suggestion
`FaaSDetailedExceptions` occur when the request has been rejected by the service. For instance, if the UUID of the lambda for invocation request does not exist.
```

> In order to get more details we recommend using the `getFaaSError`-method. A detailed list of possible errors can be found [here](liveperson-functions-foundations-error-codes.html). `FaaSLambdaException` inherits from `FaaSDetailedException`. It is only raised during invocations, if the error is caused by the implementation of the function itself.

Reviewer: ```suggestion
In order to get more details, we recommend using the `getFaaSError`-method. A detailed list of possible errors can be found [here](liveperson-functions-foundations-error-codes.html). `FaaSLambdaException` inherits from `FaaSDetailedException`. It is only raised during invocations if the error is caused by the implementation of the function itself.
```

For instance if the function returns an error on purpose or the function has a timeout. We recommend to monitor all the exceptions by using e.printStackTrace(), as this provides way more details then the error message alone. Alerting should only be done for `FaaSExceptions` or `FaaSDetailedException`.

#### JavaScript Client

Our client is available within NPM. You can head over [here](https://www.npmjs.com/package/liveperson-functions-client) to get the latest version. Than you can leverage either npm or yarn to install it as a dependency:

```bash
$ yarn add @liveperson/functions-client

$ npm install @liveperson/functions-client
```

{:.notice}

> The default Authentication method is OAuth 2.0 with grant type `client_credentials`. Please check the [above chapter](#authorization-process) on the necessary steps

Reviewer: ```suggestion
The default Authentication method is OAuth 2.0 with grant type `client_credentials`. Please check the [above chapter](#authorization-process) on the necessary steps.
```

You can leverage the following method that are exposed by the client:

* `invoke` this method allows to invoke a productive function by it's UUID. Or all productive function listening to an specific event.

> * `isImplemented` this method allows to check if an event or a specific uuid is implemented. It is a best practice to check this prior to invoking

Reviewer: we should check if is correct " allows to check " (for me sounds good but is not clear, maybe some native English speaker can help) or is "allows checking"

> * `getLambdas` this method will return all functions and there states that belong to specified account

Reviewer: ```suggestion
* `getLambdas` this method will return all functions and their states that belong to a specified account
```

{:.important}

> A more detailed version of the documentation can be found inside our [github repository](https://github.com/LivePersonInc/faas-client-node). If you experience an issue or have a wish, please create a github issue.

Reviewer: Think should be GitHub instead of github, also Grammarly says that
